src/integration.ts

```diff
@@ -301,20 +302,143 @@ export function makeIntegration<K extends string, F extends AnyFunction>(
+ * may resolve to at runtime.
+ *
+ * ```ts
+ * declare const table1;
+ * declare const table2;
+ *
+ * const tables = [table1, table2];
+ *
+ * const a = table1;
+ *    // ^ [table1]
+ *
+ * for (const a of tables) {
+ *   const b = a;
+ *          // ^ [table1, table2]
+ *
+ *   const { appsync: { getItem } } = a;
+ *                     // ^ [ table1.appsync.getItem, table2.appsync.getItem ]
+ * }
+ *
+ * const a = tables[0];
+ *          // ^ [table1]
+ *
+ * const { appsync: { getItem } } = table[0];
+ *                   // ^ [ table1.appsync.getItem ]
+ *
+ * const { appsync: { getItem = table1.appsync.getItem } } = table[2];
+ *                   // ^ [ table1.appsync.getItem ] (because of initializer)
+ * ```
+ *
+ * @param node the node to resolve the possible {@link Integration}s of.
+ * @returns a list of all the {@link Integration}s that the {@link node} could evaluate to.
+ */
+export function tryFindIntegrations(node: FunctionlessNode): Integration[] {
+  return resolve(node, undefined).filter(isIntegration);
+
+  /**
+   * Resolve all of the possible values that {@link node} may resolve to at runtime.
+   *
+   * @param node
+   * @param defaultValue default value to use if the value cannot be resolved (set by default initializers in BindingElement)
+   * @returns an array of all the values the {@link node} resolves to.
+   */
+  function resolve(
+    node: FunctionlessNode | undefined,
+    defaultValue: FunctionlessNode | undefined
+  ): any[] {
+    if (node === undefined) {
+      if (defaultValue === undefined) {
+        return [];
+      } else {
+        return resolve(defaultValue, undefined);
+      }
+    } else if (isReferenceExpr(node) || isThisExpr(node)) {
+      return [node.ref()];
+    } else if (isIdentifier(node)) {
+      return resolve(node.lookup(), defaultValue);
+    } else if (isBindingElem(node)) {
+      return resolve(node.parent, node.initializer).flatMap((pattern) => {
+        if (isIdentifier(node.name)) {
+          return [pattern[node.name.name]];
+        } else {
+          throw new Error("should be impossible");
+        }
+      });
+    } else if (isBindingPattern(node)) {
+      // we only ever evaluate `{ a }` or `[ a ]` when walking backwards from `a`
+      // the BindingElem resolver case will pluck `a` from the object returned by this
+      return resolve(node.parent, defaultValue);
+    } else if (isVariableDecl(node)) {
+      return resolve(node.initializer, defaultValue);
+    } else if (isPropAccessExpr(node) || isElementAccessExpr(node)) {
+      return resolve(node.expr, undefined).flatMap((expr) => {
+        const key = isPropAccessExpr(node)
+          ? node.name.name
+          : isStringLiteralExpr(node.element) ||
+            isNumberLiteralExpr(node.element)
+          ? node.element.value
+          : undefined;
+        if (key !== undefined) {
+          return [(<any>expr)?.[key]];
+        }
+        return [];
+      });
```

Comment:
`evalToConstant` to try to get things like template expressions and string concat?